### PR TITLE
Allow PSP's to specify a whitelist of allowed paths for host volume

### LINKED ISSUE #29326: Pods should not be able to mount arbitrary host volumes
Kubernetes administrators should be able to constrain through policies which (if any) host volumes are mountable by a pod.  Otherwise, pods can obtain access to sensitive data on the host (certificates and so forth) that an administrator might not want them to have access to.

### PR DESCRIPTION
**Pull request by jhorwit2**
**What this PR does / why we need it**:

This PR adds the ability to whitelist paths for the host volume to ensure pods cannot access directories they aren't supposed to. E.g. `/var/lib/kubelet`, `/etc/kubernetes/*`, etc. 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #29326

**Special notes for your reviewer**:

**Release note**:

### CODE REVIEW

File: pkg/security/podsecuritypolicy/provider.go [hunk 1]
```diff
@@ -329,3 +338,20 @@ func (s *simpleProvider) isValidHostPort(port int) bool {
...
+		return nil
+	}
+
+	for _, allowedPath := range allowedPaths {
+		if strings.HasPrefix(hostPath, allowedPath) {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("Host path %s is not allowed to be used. Allowed host paths: %v", hostPath, allowedPaths)
```
Comment (php-coder) [inline pkg/security/podsecuritypolicy/provider.go]: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

**Comment by php-coder** [comment]
It's a bit confusing that release note and PSP member imply path names but algorithm working like path prefix. I think it's counter-intuitive. Could we improve this?

Also I would better to use strict comparison by default. I could imagine situation when dir contains two project dirs named `foo` and `foobar` and we want to restrict access only to `foo` dir. With the current approach it's impossible because it also deny access to `foobar`.

What do you think about adding support for special `*` character at the end for a prefix behavior? IMHO it also emphasizes a fact that we're restricting by wildcard.

**Comment by php-coder** [comment]
@mfischer-zd (as initial requester) what do you think about suggested behavior? Is it ok?

**Comment by jhorwit2** [comment]
I was going back and forth on strict vs prefix. I didn't think about the example you said (`/foo` vs `/foobar`.

With `*`, would you want to support `/foo/bar/*/baz`?

**Comment by mfischer-zd** [comment]
My own view is that prefixes that look like mount points (with an implicit `/` suffix to prevent the `/foo` matching `/foobar` problem, but without an explicit `*` suffix) are sufficient.  I've never personally needed any other kind of matching for bind mount whitelists.

**Review by pweil- — COMMENTED** [review]
I like where this is going.  In addition to the question posed by @php-coder on syntax:

Should we validate against the host plugin for this to be set and throw errors if not or does it make sense to allow this to exist even with no host volume in `Volumes`?

Using an attached slice like this means you need a permutation of the PSP every time you want to give some (but not others) access to a different list of host volumes.  Is there a way to not have that coupling?  Would you just recommend namespace level PSPs at that point?

**Comment by jhorwit2** [comment]
@pweil- I feel like multiple permutations of PSP's are going to be common. That's already the case with host ports. The reason I want this feature is to restrict pods based on service accounts & namespaces to specific host volume paths. In that scenario i'd have a PSP per SA/namespace.

**Comment by pweil-** [comment]
> I feel like multiple permutations of PSP's are going to be common.

Agree.  Very similar to roles IMO, just a pain point I've been wary of. 

> In that scenario i'd have a PSP per SA/namespace.

Fair.

**Comment by pweil-** [comment]
@kubernetes/api-reviewers @liggitt @erictune @timstclair

**Comment by liggitt** [comment]
cc @kubernetes/sig-auth-api-reviews

**Comment by smarterclayton** [comment]
Rbac has path rules, can we make these consistent with those safely?

**Comment by smarterclayton** [comment]
On an unrelated note we should have called this ClusterPodSecurityPolicy so we could also have a namespace scoped rule that further confines (so you could have access to a wider scope and then tighten).  We should look at discussing what use cases for namespace scope need to be solved in a separate issue

**Comment by jhorwit2** [comment]
@smarterclayton there was a recent PR to make PSP's have the option to be namespace scoped.

**Comment by smarterclayton** [comment]
Link?  Would be good to update the PSP proposal first if so

**Comment by jhorwit2** [comment]
https://github.com/kubernetes/kubernetes/pull/42360

**Comment by liggitt** [comment]
> I could imagine situation when dir contains two project dirs named foo and foobar and we want to restrict access only to foo dir.

I would not expect to deny access to specific directories, but just to enable a whitelist mode.

I *think* allowing all subpaths of any whitelisted paths automatically makes sense, but definition need to require them to be subpaths, not just suffixes

File: pkg/security/podsecuritypolicy/provider.go [hunk 2]
```diff
@@ -329,3 +338,20 @@ func (s *simpleProvider) isValidHostPort(port int) bool {
...
+// isHostPathAllowed returns no error if the host path is a prefix of any of the
+// allowed host paths allowed by the PSP.
+func isHostPathAllowed(hostPath string, allowedPaths []string) error {
+	// If no allowed paths are specified then allow any path
+	if len(allowedPaths) == 0 {
+		return nil
+	}
+
+	for _, allowedPath := range allowedPaths {
+		if strings.HasPrefix(hostPath, allowedPath) {
```
Comment (liggitt) [inline pkg/security/podsecuritypolicy/provider.go]: this logic needs to check for exact matches or prefix matches with a trailing `/`

**Comment by jhorwit2** [comment]
@liggitt so if i'm understanding you correctly:

Ensure both paths have the trailing slash, then call `HasPrefix`. That checks if the paths are equal or if the host path is a subdir of the allowed path.

| Host Path        | Allowed Path           | approved  | why |
| ------------- |-------------|-----|-----|
| /foo      | /foo | yes | `/foo/` == `/foo/`
| /foo/bar      | /foo     |   yes | `/foo/bar/` is subdir of `/foo/`
| /foo |  /foo/bar   |    no  |  `/foo/` is not subdir or equal to `/foo/bar/`
| /foobar |  /foo   |    no  | `/foobar/` is not subdir or equal to `/foo/`

**Comment by liggitt** [comment]
yes, see https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/util.go#L121-L144 for non-allocating implementation

**Comment by jhorwit2** [comment]
Suggestions on making that public vs moving to a better util file vs copy/pasting?

**Comment by mfischer-zd** [comment]
@jhorwit2 Your example would meet my behavioral expectations :)

**Comment by liggitt** [comment]
> Suggestions on making that public vs moving to a better util file vs copy/pasting?

I'd either copy it or move it to a util package under k8s.io/apimachinery or k8s.io/apiserver trees.

Be aware that the existing function wouldn't allow `/foo` if `/foo/` was in the whitelist. Since normalizing to ensure a trailing `/` allocates, I'd probably normalize both to remove trailing `/` instead (which subslices, but does not allocate).

**Comment by grodrigues3** [comment]
@k8s-bot ok to test

**Comment by jhorwit2** [comment]
Is it normal to have a bunch of files change when you run `hack/update-staging-client-go.sh`? It appears every `BUILD` file in the staging client-go got updated?

**Comment by jhorwit2** [comment]
@k8s-bot kops aws e2e test this

**Comment by jhorwit2** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by thockin** [comment]
This reads as the opposite of starting loose and getting tighter - it's a
whitelist, no?  It's and entitlement, not a restriction.

**Comment by jhorwit2** [comment]
@thockin the default is to allow any host volume path, so it does start loose then get tighter once you specify allowed paths.

**Comment by thockin** [comment]
It starts loose then gets really tight then loosens?

**Comment by jhorwit2** [comment]
@liggitt is the latest update what you had in mind for checking?

@thockin Why do you think the security policy loosens again? First off, you need to enable host path volumes, so you are still secure by default. Once you decide you want to enable host volumes for a given PSP, you can use this feature to allow only certain paths and not every path like it currently does.

**Comment by thockin** [comment]
Maybe I misunderstand.  I assumed:

today's behavior:
- if you can't use hostPath, no problem
- if you can use hostPath, you can hostPath anything

proposed behavior:
- if you can't use hostPath, no problem
- if you can use hostPath, you can hostPath anything
- unless the whitelist is specified, in which case you can only hit things
in that list


If the proposal is actually:
- if you can't use hostPath, no problem
- if you can use hostPath, you can only hit things in the whitelist

Isn't that a breaking change?

**Comment by jhorwit2** [comment]
> proposed behavior:
> - if you can't use hostPath, no problem
> - if you can use hostPath, you can hostPath anything
> - unless the whitelist is specified, in which case you can only hit things
in that list

is correct

I think we might have a disconnect. Would you mind explaining what you mean by 
> It starts loose then gets really tight then loosens?


Also, what your problem with the proposed way is, if any?

Would you prefer it be more like how host network & host ports work?

**Comment by thockin** [comment]
On Mon, Apr 24, 2017 at 10:18 AM, Josh Horwitz <notifications@github.com> wrote:
>
> proposed behavior:
>
> if you can't use hostPath, no problem
> if you can use hostPath, you can hostPath anything
> unless the whitelist is specified, in which case you can only hit things
> in that list
>
> is correct
>
> I think we might have a disconnect. Would you mind explaining what you mean by
>
>> It starts loose then gets really tight then loosens?

sure:

> if you can use hostPath, you can hostPath anything

loose

> unless the whitelist is specified

very tight

> in which case you can only hit things in that list

loosening as whitelists are added

> Also, what your problem with the proposed way is, if any?

I don't think I have a problem, per se, just pointing out that it is different

**Comment by jhorwit2** [comment]
Thoughts on making this like HostNetwork & HostPorts?

Change `AllowedHostVolumes` to `HostVolumes`:
    * Default is nothing (empty list). 
    * All host paths would be `[*]`
    * Specify allowed prefixes same as current PR. 

Question: would this possible to do without breaking backwards compat? I.e. give all existing psp's [*] & all new ones []?

**Comment by liggitt** [comment]
> would this possible to do without breaking backwards compat? I.e. give all existing psp's [*] & all new ones []?

we *really* don't want to add fields where we have to distinguish between nil and []

**Comment by thockin** [comment]
yeah, please let's not make nil and [] significant.

File: pkg/security/podsecuritypolicy/util/util.go [hunk 3]
```diff
@@ -162,3 +163,64 @@ func PSPAllowsFSType(psp *extensions.PodSecurityPolicy, fsType extensions.FSType
...
+		return false
+	}
+
+	pathPrefixLength := len(pathPrefix)
+
+	if len(s) == pathPrefixLength {
+		// Exact match
+		return true
+	}
+	if strings.HasSuffix(pathPrefix, "/") {
```
Comment (ericchiang) [inline pkg/security/podsecuritypolicy/util/util.go]: Isn't the trailing slash already removed via "normalizePath"? When would this ever match?

File: pkg/security/podsecuritypolicy/util/util.go [hunk 4]
```diff
@@ -162,3 +163,64 @@ func PSPAllowsFSType(psp *extensions.PodSecurityPolicy, fsType extensions.FSType
...
+		// Check this instead of normalizing pathPrefix to avoid allocating on every call
+		return true
+	}
+
+	return false
+}
+
+// normalizePath returns a subslice of the original string without a trailing slash.
+// if no trailing slash is found the original string is returned
+func normalizePath(path string) string {
```
Comment (ericchiang) [inline pkg/security/podsecuritypolicy/util/util.go]: This is just `strings.TrimSuffix(path, "/")`

**Comment by jhorwit2** [comment]
@ericchiang @liggitt what are the next steps? I addressed the comments, but not sure where to go from here other than to keep fixing rebase issues.

**Comment by ericchiang** [comment]
I'd like to see this in for 1.7. @liggitt @thockin it looks like there aren't any outstanding concerns. Can we get this merged?

**Comment by smarterclayton** [comment]
I'd also like to see this in 1.7.  Let's try to resolve any API concerns on
monday.

**Comment by jhorwit2** [comment]
@smarterclayton do you have any api concerns?

**Comment by smarterclayton** [comment]
LGTM.  Rebase and I'll tag you.

**Comment by jhorwit2** [comment]
@k8s-bot pull-kubernetes-verify test this

**Comment by smarterclayton** [comment]
Need to regenerate docs.

**Comment by jhorwit2** [comment]
@smarterclayton yeah, i'm running into issues. I can run the verify commands locally and all is fine. If I run the generate commands there are no changes to commit... :/

**Comment by smarterclayton** [comment]
You ran hack/update-api-reference-docs.sh?

On May 27, 2017, at 9:31 AM, Josh Horwitz <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> yeah, i'm running into
issues. I can run the verify commands locally and all is fine. If I run the
generate commands there are no changes to commit... :/

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/43946#issuecomment-304451770>,
or mute the thread

**Comment by jhorwit2** [comment]
@smarterclayton so I can run that; however, the recommended command to before that, `hack/update-swagger-spec.sh`, fails 100% of the time on the autoscaling/v2alpha1 resource... Which is strange because I didn't modify anything related to that api group. :(

**Comment by smarterclayton** [comment]
Works for me.  Run make clean, then remove the `.make` directory and try
again.

When I run update-swagger-spec on your branch it adds the new field to the
doc.

**Comment by jhorwit2** [comment]
@smarterclayton thanks for trying it out. I ended up setting up a fresh ubuntu vm which worked. I guess something is up with my mac that's causing issues.

**Comment by jhorwit2** [comment]
@smarterclayton rebase and tests look good. 👍

**Comment by liggitt** [comment]
removed release note pending revert of the API change in https://github.com/kubernetes/kubernetes/pull/47851

**Pull request merged**